### PR TITLE
[Perf] Trim HunyuanVideo encoder padding tokens

### PR DESCRIPTION
**Pull request by david6666666**
## Summary

This PR trims tail padding from HunyuanVideo-1.5 text encoder tokens on the single-GPU, non-sequence-parallel path.

After prompt token reordering, valid encoder tokens are already compacted to the front of `encoder_hidden_states` / `encoder_attention_mask`. The transformer blocks were still receiving the padded tail. This patch slices both tensors to the max valid encoder length in the batch before the repeated transformer blocks. If the resulting mask is all true, it passes `None` instead of an all-true mask.

The change is guarded by `if not ctx.sp_active`, so sequence-parallel execution keeps the existing path unchanged.

## Why This Helps

The text encoder padding is paid repeatedly by every HunyuanVideo transformer block. Trimming it reduces the K/V sequence length for cross-attention and removes redundant mask work. The benefit shows up with both `CUDNN_ATTN` and `TORCH_SDPA` because both backends avoid computing attention over padded encoder tokens after the trim.

## B300 Results

Model: `hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v`

Common settings: prompt `A cat walks on the grass, realistic style.`, `height=480`, `width=832`, `num_frames=33`, `num_inference_steps=50`, `guidance_scale=6.0`, `flow_shift=5.0`, `fps=24`, `seed=42`.

| Backend / topology | Baseline total | PR total | Total speedup | Baseline forward | PR forward | Forward speedup | Precision check |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | --- |
| `CUDNN_ATTN`, 1 GPU | 27.4157s | 16.8355s | 38.6% | 27.0295s | 16.4482s | 39.1% | SSIM 0.694584, PSNR 22.200918 |
| `TORCH_SDPA`, 1 GPU | 99.1596s | 31.2140s | 68.5% | 98.771831s | 30.827799s | 68.8% | SSIM 0.822205, PSNR 24.767672 |
| `CUDNN_ATTN`, 2 GPU SP (`--ulysses-degree 2`) | 20.1943s | 20.1452s | 0.2% | 19.825956s | 19.796866s | 0.1% | No material effect expected; patch is guarded out under SP |

CUDNN baseline self-run reference on the same setup: SSIM 0.702010, PSNR 22.064819. The CUDNN optimized-vs-baseline metrics are within the observed same-seed CUDNN run-to-run variance.

## E2E Reproduction

Single GPU CUDNN or SDPA:

```bash
CUDA_VISIBLE_DEVICES=3 \
PYTHONPATH=$PWD \
DIFFUSION_ATTENTION_BACKEND=CUDNN_ATTN \
/home/zjy/code/david/.venv/bin/python \
examples/offline_inference/text_to_video/text_to_video.py \
  --model hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v \
  --prompt 'A cat walks on the grass, realistic style.' \
  --height 480 \
  --width 832 \
  --num-frames 33 \
  --num-inference-steps 50 \
  --guidance-scale 6.0 \
  --flow-shift 5.0 \
  --fps 24 \
  --seed 42 \
  --output /tmp/hv15_50step.mp4 \
  --enable-diffusion-pipeline-profiler
```

For SDPA, change `DIFFUSION_ATTENTION_BACKEND=CUDNN_ATTN` to `DIFFUSION_ATTENTION_BACKEND=TORCH_SDPA`.

Two-GPU sequence parallel check:

```bash
CUDA_VISIBLE_DEVICES=2,3 \
PYTHONPATH=$PWD \
DIFFUSION_ATTENTION_BACKEND=CUDNN_ATTN \
/home/zjy/code/david/.venv/bin/python \
examples/offline_inference/text_to_video/text_to_video.py \
  --model hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v \
  --prompt 'A cat walks on the grass, realistic style.' \
  --height 480 \
  --width 832 \
  --num-frames 33 \
  --num-inference-steps 50 \
  --guidance-scale 6.0 \
  --flow-shift 5.0 \
  --fps 24 \
  --seed 42 \
  --ulysses-degree 2 \
  --output /tmp/hv15_sp2_50step.mp4 \
  --enable-diffusion-pipeline-profiler
```

Precision metrics:

```bash
ffmpeg -hide_banner -i baseline.mp4 -i optimized.mp4 -lavfi "[0:v][1:v]ssim" -f null -
ffmpeg -hide_banner -i baseline.mp4 -i optimized.mp4 -lavfi "[0:v][1:v]psnr" -f null -
```

## Validation

- `python3 -m py_compile vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py`
- B300 E2E 50-step single-GPU benchmark with `CUDNN_ATTN`
- B300 E2E 50-step single-GPU benchmark with `TORCH_SDPA`
- B300 E2E 50-step two-GPU sequence-parallel benchmark with `CUDNN_ATTN`
- FFmpeg SSIM/PSNR comparison for generated videos

### CODE REVIEW

**Comment by david6666666** [comment]
Additional B300 validation update after trimming the PR down to only `vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py`.

## Scope cleanup

The branch was rebased/reset onto current `upstream/main` and force-pushed. The PR now contains one signed commit and one changed file only:

- `vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py`

## TORCH_SDPA single-GPU E2E

Model: `hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v`

Settings: `CUDA_VISIBLE_DEVICES=3`, `DIFFUSION_ATTENTION_BACKEND=TORCH_SDPA`, prompt `A cat walks on the grass, realistic style.`, `height=480`, `width=832`, `num_frames=33`, `seed=42`.

| Steps | Variant | Total time | Forward time |
| ---: | --- | ---: | ---: |
| 4 | baseline | 9.2456s | 8.865622s |
| 4 | this PR | 3.7839s | 3.425235s |
| 50 | baseline | 99.1596s | 98.771831s |
| 50 | this PR | 31.2140s | 30.827799s |

50-step speedup:

- Total: 68.5%
- Forward: 68.8%

FFmpeg precision comparison, 50-step optimized vs baseline:

- SSIM All: 0.822205
- PSNR average: 24.767672

## CUDNN_ATTN two-GPU sequence-parallel E2E

Settings: `CUDA_VISIBLE_DEVICES=2,3`, `DIFFUSION_ATTENTION_BACKEND=CUDNN_ATTN`, `--ulysses-degree 2`, same model/prompt/resolution/frame/seed settings.

| Steps | Variant | Total time | Forward time |
| ---: | --- | ---: | ---: |
| 4 | baseline | 3.5998s | 3.260937s |
| 4 | this PR | 3.5854s | 3.243092s |
| 50 | baseline | 20.1943s | 19.825956s |
| 50 | this PR | 20.1452s | 19.796866s |

Conclusion: no material SP impact, as expected. The patch is guarded by `if not ctx.sp_active`, so sequence-parallel execution keeps the original encoder token path.

## CUDNN_ATTN single-GPU reference

Previously measured 50-step B300 single-GPU result:

| Variant | Total time | Forward time |
| --- | ---: | ---: |
| baseline | 27.4157s | 27.0295s |
| this PR | 16.8355s | 16.4482s |

Speedup:

- Total: 38.6%
- Forward: 39.1%

Precision:

- Optimized vs baseline: SSIM All 0.694584, PSNR average 22.200918
- Baseline self-run reference: SSIM All 0.702010, PSNR average 22.064819

## Repro commands

Single GPU CUDNN/SDPA:

```bash
CUDA_VISIBLE_DEVICES=3 \
PYTHONPATH=$PWD \
DIFFUSION_ATTENTION_BACKEND=TORCH_SDPA \
/home/zjy/code/david/.venv/bin/python \
examples/offline_inference/text_to_video/text_to_video.py \
  --model hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v \
  --prompt 'A cat walks on the grass, realistic style.' \
  --height 480 \
  --width 832 \
  --num-frames 33 \
  --num-inference-steps 50 \
  --guidance-scale 6.0 \
  --flow-shift 5.0 \
  --fps 24 \
  --seed 42 \
  --output /tmp/hv15_sdpa_50step.mp4 \
  --enable-diffusion-pipeline-profiler
```

SP check:

```bash
CUDA_VISIBLE_DEVICES=2,3 \
PYTHONPATH=$PWD \
DIFFUSION_ATTENTION_BACKEND=CUDNN_ATTN \
/home/zjy/code/david/.venv/bin/python \
examples/offline_inference/text_to_video/text_to_video.py \
  --model hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v \
  --prompt 'A cat walks on the grass, realistic style.' \
  --height 480 \
  --width 832 \
  --num-frames 33 \
  --num-inference-steps 50 \
  --guidance-scale 6.0 \
  --flow-shift 5.0 \
  --fps 24 \
  --seed 42 \
  --ulysses-degree 2 \
  --output /tmp/hv15_sp2_50step.mp4 \
  --enable-diffusion-pipeline-profiler
```

Metrics:

```bash
ffmpeg -hide_banner -i baseline.mp4 -i optimized.mp4 -lavfi "[0:v][1:v]ssim" -f null -
ffmpeg -hide_banner -i baseline.mp4 -i optimized.mp4 -lavfi "[0:v][1:v]psnr" -f null -
```

**Comment by david6666666** [comment]
baseline:

https://github.com/user-attachments/assets/2254552b-76f4-4873-9c11-8811a6c26485


opt:

https://github.com/user-attachments/assets/0ac8184d-44e6-4f27-9d45-53ac6a4da229

**Comment by david6666666** [comment]
Additional HunyuanVideo-1.5 I2V validation for this encoder padding trim.

## Setup

Model: `hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_i2v`

Runtime/settings:

- B300, `CUDA_VISIBLE_DEVICES=3`
- `DIFFUSION_ATTENTION_BACKEND=TORCH_SDPA`
- input image: `cherry_blossom.jpg` from the offline I2V example README
- prompt: `Cherry blossoms swaying gently in the breeze, petals falling, smooth motion`
- `height=480`, `width=832`, `num_frames=33`, `num_inference_steps=50`
- `guidance_scale=6.0`, `flow_shift=5.0`, `fps=24`, `seed=42`

Note: I attempted `CUDNN_ATTN` on current `upstream/main`, but current main does not include that backend (`Valid options are: FLASH_ATTN, TORCH_SDPA, SAGE_ATTN`). Since this PR is now intentionally trimmed to only the HunyuanVideo transformer source change, I used `TORCH_SDPA` for the clean upstream-main baseline comparison.

## E2E result

| Variant | Total generation time | Pipeline forward time |
| --- | ---: | ---: |
| upstream/main baseline | 99.2804s | 98.923084s |
| this PR | 34.5925s | 34.234433s |

Speedup:

- Total time: 65.2%
- Forward time: 65.4%
- Ratio: ~2.87x total / ~2.89x forward

## Video comparison

FFmpeg optimized-vs-baseline metrics:

- SSIM All: 0.922008
- PSNR average: 29.191279

Both outputs are 832x480, 33 frames, 24 fps.

## Repro command

```bash
CUDA_VISIBLE_DEVICES=3 \
PYTHONPATH=$PWD \
DIFFUSION_ATTENTION_BACKEND=TORCH_SDPA \
/home/zjy/code/david/.venv/bin/python \
examples/offline_inference/image_to_video/image_to_video.py \
  --model hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_i2v \
  --image /home/zjy/code/david/bench_inputs/cherry_blossom.jpg \
  --prompt 'Cherry blossoms swaying gently in the breeze, petals falling, smooth motion' \
  --height 480 \
  --width 832 \
  --num-frames 33 \
  --num-inference-steps 50 \
  --guidance-scale 6.0 \
  --flow-shift 5.0 \
  --fps 24 \
  --seed 42 \
  --output /tmp/hv15_i2v_sdpa_50step_33f.mp4 \
  --enable-diffusion-pipeline-profiler
```

Metrics:

```bash
ffmpeg -hide_banner -i baseline.mp4 -i optimized.mp4 -lavfi "[0:v][1:v]ssim" -f null -
ffmpeg -hide_banner -i baseline.mp4 -i optimized.mp4 -lavfi "[0:v][1:v]psnr" -f null -
```

**Comment by david6666666** [comment]
Additional clarification on the normal single-GPU path.

For the normal single-GPU, non-sequence-parallel path, the new code takes the following branches:

```python
ctx = get_forward_context()
if not ctx.sp_active:
    max_valid_encoder_tokens = int(encoder_attention_mask.sum(dim=1).max().item())
    if max_valid_encoder_tokens < encoder_attention_mask.shape[1]:
        encoder_hidden_states = encoder_hidden_states[:, :max_valid_encoder_tokens]
        encoder_attention_mask = encoder_attention_mask[:, :max_valid_encoder_tokens]
    if encoder_attention_mask.all():
        encoder_attention_mask = None
```

Expected branch behavior:

1. `ctx.sp_active == False`, so the optimization block is entered.
2. `max_valid_encoder_tokens` is computed from the already-reordered encoder mask.
3. In the common HunyuanVideo-1.5 case, `max_valid_encoder_tokens < encoder_attention_mask.shape[1]` is true because the reordered encoder sequence is `[valid_image, valid_byte5, valid_mllm, padding]` and still contains tail padding.
4. The code slices both `encoder_hidden_states` and `encoder_attention_mask` to the max valid encoder token length in the batch.
5. After this trim, `encoder_attention_mask.all()` is typically true because the remaining tokens are all valid. In that case, the mask is set to `None`.

So the typical single-GPU result is:

```text
encoder_hidden_states: [B, original_len, C] -> [B, max_valid_len, C]
encoder_attention_mask: [B, original_len] -> None
```

This is the source of the speedup: every following HunyuanVideo transformer block avoids processing tail-padding encoder tokens, and the attention path also avoids carrying an all-true encoder mask.

**Comment by david6666666** [comment]
@SamitHuang @gcanlin ptal thx

**Comment by david6666666** [comment]
@RuixiangMa @lishunyang12 ptal thx

**Review by lishunyang12 — APPROVED** [review]

File: vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py [hunk 1]
```diff
@@ -764,6 +764,13 @@ def forward(
 
         # Create explicit attn_mask for image tokens when SP auto_pad is active.
         ctx = get_forward_context()
+        if not ctx.sp_active:
+            max_valid_encoder_tokens = int(encoder_attention_mask.sum(dim=1).max().item())
+            if max_valid_encoder_tokens < encoder_attention_mask.shape[1]:
+                encoder_hidden_states = encoder_hidden_states[:, :max_valid_encoder_tokens]
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/hunyuan_video/hunyuan_video_15_transformer.py:770]: Padding is already masked to `-inf`, so trimming it should be a numerical no-op. That means SSIM 0.69 vs the 0.70 baseline self-run is within CUDNN run-to-run noise and doesn't really confirm equivalence. Can you compare the pre-VAE latent (rel L2 / cosine) under deterministic kernels instead?

**Comment by david6666666** [comment]
Follow-up on the pre-VAE latent equivalence check.

I ran a same-process / same-loaded-model A/B test to avoid the run-to-run noise we saw from separate E2E processes. The test fixes the prompt, prompt embeddings, initial latent tensor, BF16 dtype, TF32 off, deterministic flags, and uses `TORCH_SDPA` math-only kernels. Then it directly compares transformer forward outputs for three variants:

1. `trim_on`: current PR behavior
2. `trim_off`: temporary local monkeypatch disabling the trim branch
3. `trim_keep_mask`: trim enabled, but keeping the all-true encoder attention mask instead of setting it to `None`

Results for `hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v`, 480p, 33 frames, 50-step schedule:

| timestep | positive rel L2 | positive cosine | positive MAE | positive MSE | max abs |
|---|---:|---:|---:|---:|---:|
| `t0=1000.0` | `0.0058497` | `0.99998298` | `0.0050655` | `4.85e-05` | `0.03515625` |
| `t25=833.3333` | `0.0058615` | `0.99998283` | `0.0045029` | `3.83e-05` | `0.03125` |
| `t49=92.5926` | `0.0061017` | `0.99998139` | `0.0047543` | `4.21e-05` | `0.041015625` |

The negative CFG branch shows the same pattern: rel L2 around `0.00527 - 0.00554`, cosine around `0.9999846 - 0.9999861`.

Important split result:

- `trim_on` vs `trim_keep_mask` is bit-exact: `rel_l2=0.0`, `cosine=1.0`, `exact_equal=true`.
- Therefore, the `encoder_attention_mask = None` part is not causing the difference.
- The observed non-bit-exact difference comes from removing the masked padding tokens and shortening the attention sequence. Mathematically this should be equivalent because those tokens are masked, but numerically the BF16 attention path changes due to the different sequence length / reduction path.

So I would describe this as numerically very close, but not bit-exact. The direct transformer-forward cosine is consistently `>0.99998` under deterministic settings.

**Review by gcanlin — APPROVED** [review]
